third_party/bigframes_vendored/ibis/backends/sql/compilers/base.py

```diff
@@ -0,0 +1,1660 @@
+# Contains code from https://github.com/ibis-project/ibis/blob/main/ibis/backends/sql/compilers/base.py
+
+from __future__ import annotations
+
+import abc
+import calendar
+from functools import partial, reduce
+import itertools
+import math
+import operator
+import string
+from typing import Any, ClassVar, TYPE_CHECKING
+
+from ibis.backends.sql.rewrites import (
+    add_one_to_nth_value_input,
+    add_order_by_to_empty_ranking_window_functions,
+    empty_in_values_right_side,
+    FirstValue,
+    LastValue,
+    lower_bucket,
+    lower_capitalize,
+    lower_sample,
+    one_to_zero_index,
+    sqlize,
+)
+import ibis.common.exceptions as com
+import ibis.common.patterns as pats
+from ibis.config import options
+import ibis.expr.datatypes as dt
+import ibis.expr.operations as ops
+from ibis.expr.operations.udf import InputType
+from ibis.expr.rewrites import lower_stringslice
```

Comment:
Should we change this to our vendored copy?

Reply:
Fixed in this PR. Thanks!